fix(types): replace schema-based path tracking with CurrentPath parameter

kosei28 opened the pull request:
- ref: #4529

### The author should do the following, if applicable

- [x] Add tests
- [x] Run tests
- [x] `bun run format:fix && bun run lint:fix` to format the code
- [ ] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

kosei28 commented:
The number of instantiations is now 694,649.

kosei28 commented:
The number of instantiations for the `CurrentPath` approach is 684,676.

yusukebe commented on src/types.ts:
```diff
@@ -2069,25 +2262,50 @@ export interface OnHandlerInterface<
+  Merged,
+] extends [Promise<void>]
+  ? VoidMiddlewarePath | Path
+  : [Merged] extends [never]
```
Are these lines necessary? This means you can apply the following patch:

```diff
diff --git a/src/types.ts b/src/types.ts
index 27aeefaa..5766b474 100644
--- a/src/types.ts
+++ b/src/types.ts
@@ -2303,9 +2303,7 @@ type AddVoidMiddlewarePath<Merged, VoidMiddlewarePath extends string, Path exten
   Merged,
 ] extends [Promise<void>]
   ? VoidMiddlewarePath | Path
-  : [Merged] extends [never]
-    ? VoidMiddlewarePath | Path
-    : VoidMiddlewarePath
+  : VoidMiddlewarePath
```

I think `Merged` will not be `never` in any case.

kosei28 replied:
I was just following `AddSchemaOrPath` from #4529, but yeah, `Merged` won't be `never`.

yusukebe replied:
Ah, my bad!

yusukebe commented on src/helper/factory/index.test.ts:
```diff
@@ -388,9 +388,7 @@ describe('createFactory', () => {
-      type Expected = {
-        '*': {}
-      }
+      type Expected = {}
```
Though you change the test, I also think it's okay. It's not a problem for real-world use cases. Makes sense for me 👍

kosei28 commented:
Reverted to `CurrentPath` approach and fixed type updates for pathless `app.use()` and multi-path `app.on()`.

Note: Tests pass locally, but CI is failing due to const type parameters conflicting with the old esbuild version. #4571 will fix the build.
